doc/neps/nep-0042-new-dtypes.rst

```diff
@@ -0,0 +1,1149 @@
+   This method may be expanded later (see alternatives).
+3. ``__discover_descr_from_pyobject__(cls, obj) -> dtype``: A classmethod that
+   returns the correct descriptor given the input object.
+   *Note that only parametric DTypes have to implement this*, most datatypes
+   can simply use a default (singleton) dtype instance which is found only
+   based on the ``type(obj)`` of the Python object.
+
+The ``type`` which is already associated with any dtype through the ``dtype.type``
+attribute maps the dtype to the Python type.
+However, initially we anticipate automatic conversion only for scalars which
+subclass from a base scalar object, such as ``np.generic`` (although possibly
+a class below ``np.generic`` to avoid arry-scalar behaviour). 
+This will be cached globally to create a mapping (dictionary)
+``knwon_python_types[type] = DType``.
+NumPy currently uses a small hard-coded mapping and conversion of numpy scalars
+(inheriting from ``np.generic``) to achieve this, however, this forces a certain
+structure on the associated ``type``.
+
+.. note::
+
+    Python integers do not have a clear/specific NumPy type associated to them
+    right now.  Instead they will need to be be implemented using an
+    ``AbstractPyInt``, this DType class, can than provide
+    ``__discover_descr_from_pyobject__`` and return the actual dtype which
+    is e.g. ``np.dtype("int64")``.
+    For dispatching/promotion in ufuncs, it will also probably be necessary
+    to dynamically create ``AbstractPyInt[value]`` classes (creation can be
+    cached), so that they can provide the current functionality provided by
+    ``np.result_type(python_integer, array)``.
+
+To allow for a DType to accept specific inputs as known scalars, we will
+initially use a method such as ``known_scalar_type`` or ``known_scalar_object``
+which allows discovering a e.g. ``vector`` as a scalar instead of an instance
+(if ``np.array(vector, dtype=VectorDType)`` is provided).
+This will *not* be public API initially, but may be made public at a later
+time.
+
+Once the datatype is found (a step that is skipped if the user provides it),
+the actual descriptor has to be discovered.
+For most datatypes, which are not parametric, this is always a canonical default
+instance and ``__discover_descr_from_pyobject__`` does not need to be defined.
+For parametric datatypes, however, such as strings, it is necessary to find the
+correct string length and thus inspect the given python object, this is
+also necessary for example for a Categorical datatype.
+In this case ``__discover_descr_from_pyobject__`` is called and must return
+a dtype instance.
+*Implementors should aim to avoid complex logic in this function.*
+
+After the dtype instance is found, the common dtype with the current previous
+found common dtype instance to find a single dtype that represents all inputs
+and thus support for example mixed integer and floating point input.
```

Comment:
```suggestion
After the dtype instance is found, the common dtype is compared with the current
previously-found common dtype instance to find a single dtype that represents all
inputs and thus supports, for example, mixed integer and floating point input.
```
Unless I've misunderstood, it might be more clear to say "After the dtype instance is found for each element"

Reply:
Added some pseudo-code, maybe that is a thing I generally need to really, too much text, too little pseudo-code, hmmm...

Reply:
If there was one over-arching comment I had from the readthrough, it would be this :). Perhaps the most informative part is the picture of the dtype hierarchy - if some of the text could be replaced with an example with pseudo code I think it could really improve digestibility!